brave-cxf3/src/main/java/com/github/kristofa/brave/cxf3/TracerContext.java

```diff
@@ -0,0 +1,67 @@
+package com.github.kristofa.brave.cxf3;
+
+import com.github.kristofa.brave.Brave;
+import com.github.kristofa.brave.ServerSpan;
+import com.github.kristofa.brave.ServerSpanThreadBinder;
+
+import java.util.concurrent.Callable;
+
+/**
+ * Used to handle {@link ServerSpan} retrieved from cxf message exchange.
+ */
+public class TracerContext {
+
+  private final ServerSpanThreadBinder serverSpanThreadBinder;
+  private final ServerSpan serverSpan;
+
+  TracerContext(Brave brave, ServerSpan serverSpan) {
+    this.serverSpanThreadBinder = brave.serverSpanThreadBinder();
+    this.serverSpan = serverSpan;
+  }
+
+  /**
+   * Wraps provided {@link Runnable and ensures same {@link ServerSpan} context as the thread from which the Runnable was executed.
+   * {@link ServerSpan} is taken from cxf {@link org.apache.cxf.message.Message} {@link org.apache.cxf.message.Exchange}
+   *
+   * @param command {@link Runnable}
+   * @return wrapped {@link Runnable}
+   */
+  public Runnable wrap(Runnable command) {
+    if (command == null) {
+      return null;
+    }
+
+    // we should clean this thread, ServerSpan will finish on another one
+    serverSpanThreadBinder.setCurrentSpan(null);
+
+    return () -> {
+      if (serverSpan != null) {
+        serverSpanThreadBinder.setCurrentSpan(serverSpan);
+      }
+      command.run();
+    };
+  }
+
+  /**
+   * Wraps provided {@link Callable} and ensures same {@link ServerSpan} context as the thread from which the Callable was executed.
+   * {@link ServerSpan} is taken from cxf {@link org.apache.cxf.message.Message} {@link org.apache.cxf.message.Exchange}
+   *
+   * @param command {@link Callable}
+   * @return wrapped {@link Callable}
+   */
+  public <V> Callable<V> wrap(Callable<V> command) {
+    if (command == null) {
+      return null;
+    }
+
+    // we should clean this thread, ServerSpan will finish on another one
+    serverSpanThreadBinder.setCurrentSpan(null);
```

Comment:
Yes it does.
I get it now. It should look like this, right? Also added special handling of `ServerSpan#EMPTY` in `getCurrentServerSpan`.


```java
  public Runnable wrap(Runnable command) {
    if (command == null) {
      return null;
    }

    // get server span bind to current thread
    final ServerSpan current = getCurrentServerSpan();

    return () -> {
      try {
        // When current span is not null, it means it was bind manually by developer
        // or this wrapper is used in sync scenario.
        // Bind server span from Exchange only when current span is null.
        if (current == null) {
          serverSpanThreadBinder.setCurrentSpan(serverSpan);
        }
        command.run();
      } finally {
        serverSpanThreadBinder.setCurrentSpan(current);
      }
    };
  }

  private ServerSpan getCurrentServerSpan() {
    // gets existing server span or reinitializes to ServerSpan#EMPTY using ThreadLocal#initialValue
    final ServerSpan current = serverSpanThreadBinder.getCurrentServerSpan();

    // when value was just initialized
    if (ServerSpan.EMPTY == current) {
      serverSpanThreadBinder.setCurrentSpan(null);
      return null;
    }
    return current;
  }
```

But it also means, that we still need find out way to clean server span from thread in async scenarios, 
because code fragment below (taken from `BraveServerInInterceptor`) doesn't do it's job.

```java
if (!message.getExchange().isSynchronous()) {
  threadBinder.setCurrentSpan(null);
}
```

That's why I've previously added 'BraveAsyncJaxRsThreadCleaner':

```java
class BraveAsyncJaxRsThreadCleaner implements ContainerRequestFilter {		
		
  @Context		
  private ResourceInfo resourceInfo;		
  private final ServerSpanThreadBinder serverThreadBinder;		
		
  BraveAsyncJaxRsThreadCleaner(Brave brave) {		
    this.serverThreadBinder = brave.serverSpanThreadBinder();		
  }		
		
  @Override		
  public void filter(ContainerRequestContext containerRequestContext) throws IOException {		
    if (isAsyncResponse()) {		
      serverThreadBinder.setCurrentSpan(null);		
    }		
  }		
		
  protected boolean isAsyncResponse() {		
    for (final Annotation[] annotations : resourceInfo.getResourceMethod().getParameterAnnotations()) {		
      for (final Annotation annotation : annotations) {		
        if (annotation.annotationType().equals(Suspended.class)) {		
          return true;		
        }		
      }		
    }		
    return false;		
  }		
}
```

Should I revert it's removal or do you have other idea what to do in such case?